src/software/ai/hl/stp/tactic/receiver/receiver_fsm.h

```diff
@@ -0,0 +1,296 @@
+        // when autokick is enabled
+        double kick_speed = BALL_MAX_SPEED_METERS_PER_SECOND - 1;
+        Angle shot_offset = Angle::asin(lateral_speed / kick_speed);
+
+        // check which direction the ball is going in so we can decide which direction to
+        // apply the offset in
+        if (lateral_vel.dot(shot_vector.rotate(Angle::quarter())) > 0)
+        {
+            // need to go clockwise
+            shot_offset = -shot_offset;
+        }
+        return shot_dir + shot_offset;
+    }
+
+    /**
+     * Figures out the location of the one-time shot and orientation the robot should face
+     *
+     * @param robot The robot performing the one-touch
+     * @param ball The ball on the field
+     * @param best_shot_target The point to shoot at
+     */
+    static Shot getOneTimeShotPositionAndOrientation(const Robot& robot, const Ball& ball,
+                                                     const Point& best_shot_target)
+    {
+        double dist_to_ball_in_dribbler =
+            DIST_TO_FRONT_OF_ROBOT_METERS + BALL_MAX_RADIUS_METERS;
+        Point ball_contact_point =
+            robot.position() + Vector::createFromAngle(robot.orientation())
+                                   .normalize(dist_to_ball_in_dribbler);
+
+        // Find the closest point to the ball contact point on the ball's trajectory
+        Point closest_ball_pos = ball.position();
+        if (ball.velocity().length() != 0)
+        {
+            closest_ball_pos =
+                closestPoint(ball_contact_point,
+                             Line(ball.position(), ball.position() + ball.velocity()));
+        }
+        Ray shot(closest_ball_pos, best_shot_target - closest_ball_pos);
+
+        Angle ideal_orientation      = getOneTimeShotDirection(shot, ball);
+        Vector ideal_orientation_vec = Vector::createFromAngle(ideal_orientation);
+
+        // The best position is determined such that the robot stays in the ideal
+        // orientation, but moves the shortest distance possible to put its contact
+        // point in the ball's path.
+        Point ideal_position =
+            closest_ball_pos - ideal_orientation_vec.normalize(dist_to_ball_in_dribbler);
+
+        return Shot(ideal_position, ideal_orientation);
+    }
+
+    /*
+     * Finds a shot that is greater than MIN_SHOT_NET_PERCENT_OPEN and
+     * respects MAX_DEFLECTION_FOR_ONE_TOUCH_SHOT for the highest chance
+     * of scoring with a one-touch shot. If neither of those are true, return a nullopt
+     *
+     * @param world The world to find a feasible shot on
+     * @param assigned_robot The robot that will be performing the one-touch
+     */
+    static std::optional<Shot> findFeasibleShot(const World& world,
+                                                const Robot& assigned_robot)
+    {
+        // Check if we can shoot on the enemy goal from the receiver position
+        std::optional<Shot> best_shot_opt = calcBestShotOnGoal(
```

Comment:
a quick diagram could be helpful here

Reply:
calcBestShot is currently undergoing a major overhaul by @jared-paul 

I will hold off on ASCII art here because its soon to change (and we should maybe just add it to caclBestShotOnGoal (but maybe its impossible given the complexity of drawing agnles in ascii)

Reply:
bumping